pkg/ffmpeg/codec_detector.go

```diff
@@ -0,0 +1,240 @@
+package ffmpeg
+
+import (
+	"context"
+	"regexp"
+	"strings"
+
+	"github.com/stashapp/stash/pkg/logger"
+)
+
+var HWCodecSupport []VideoCodec
+
+// Tests all (given) hardware codec's
+func FindHWCodecs(ctx context.Context, encoder FFMpeg) {
+	HWCodecSupport = HWCodecSupport[:0]
+
+	for _, codec := range []VideoCodec{
+		VideoCodecN264,
+		VideoCodecI264,
+		VideoCodecV264,
+		VideoCodecR264,
+		VideoCodecIVP9,
+		VideoCodecVVP9,
+	} {
+		var args Args
+		args = append(args, "-hide_banner")
+		args = args.LogLevel(LogLevelQuiet)
+		args = HWDeviceInit(args, codec)
+		args = args.Format("lavfi")
+		args = args.Input("color=c=red")
+		args = args.Duration(0.1)
+
+		videoFilter := HWFilterInit(codec)
+		// Test scaling
+		videoFilter = videoFilter.ScaleDimensions(-2, 160)
+		videoFilter = HWCodecFilter(videoFilter, codec)
+		args = append(args, CodecInit(codec)...)
+		args = args.VideoFilter(videoFilter)
+
+		args = args.Format("null")
+		args = args.Output("-")
+
+		cmd := encoder.Command(ctx, args)
+
+		if err := cmd.Run(); err == nil {
+			HWCodecSupport = append(HWCodecSupport, codec)
+		}
+	}
+
+	logger.Info("Supported HW codecs: ")
+	for _, codec := range HWCodecSupport {
+		logger.Info("\t", codec)
+	}
+}
+
+// Test full-hardware transcoding on an input video
+/*func HWCodecVideoSupported(ctx context.Context, encoder FFMpeg, o TranscodeStreamOptions) bool {
+	if !HWCodecDetect(o.Codec.codec) {
+		return false
+	}
+
+	var args Args
+	args = append(args, "-hide_banner")
+	args = append(args, o.ExtraInputArgs...)
+	args = args.LogLevel(LogLevelQuiet)
+	args = HWDeviceInit_Full(args, o.Codec.codec)
+	args = args.Input(o.Input)
+	args = args.Duration(0.1)
+
+	args = args.VideoCodec(o.Codec.codec)
+	if len(o.Codec.extraArgs) > 0 {
+		args = append(args, o.Codec.extraArgs...)
+	}
+
+	// Test scaling
+	videoFilter := HWFilterInit(o.Codec.codec)
+	videoFilter = videoFilter.ScaleDimensions(-2, 160)
+	videoFilter = HWCodecFilter(videoFilter, o.Codec.codec)
+	args = args.VideoFilter(videoFilter)
+
+	args = args.Format("null")
+	args = args.Output("-")
+
+	cmd := encoder.Command(ctx, args)
+
+	err := cmd.Run()
+	return err == nil
+}*/
+
+// Prepend input for hardware encoding only
+func HWDeviceInit(args Args, codec VideoCodec) Args {
```

Comment:
Unless this function (and the other functions in this file) are used outside of the package, they should probably be made private.